com-api/com-api-concept/src/lib.rs

```
@@ -160,48 +230,80 @@ where
     ///
     /// This corresponds to `MaybeUninit::write`.
     fn write(self, value: T) -> Self::SampleMut;
+
+    /// Get a mutable pointer to the internal maybe uninitialized `T`.
+    ///
+    /// The caller has to make sure to initialize the data in the buffer.
+    /// Reading from the received pointer before initialization is undefined behavior.
+    fn as_mut_ptr(&mut self) -> *mut T;
 }
 
-pub trait Interface {}
+pub trait Interface {
+    type Consumer<R: Runtime + ?Sized>: Consumer<R>;
+    type Producer<R: Runtime + ?Sized>: Producer<R>;
+}
 
-pub trait OfferedProducer {
+#[must_use = "if a service is offered it will be unoffered and dropped immediately, causing unexpected behavior in the system"]
+pub trait OfferedProducer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type Producer: Producer<Interface = Self::Interface>;
+    type Producer: Producer<R, Interface = Self::Interface>;
 
     fn unoffer(self) -> Self::Producer;
 }
 
-pub trait Producer {
+pub trait Producer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type OfferedProducer: OfferedProducer<Interface = Self::Interface>;
+    type OfferedProducer: OfferedProducer<R, Interface = Self::Interface>;
 
     fn offer(self) -> Result<Self::OfferedProducer>;
 }
 
-pub trait Consumer {}
+pub trait Publisher<T>
+where
+    T: Reloc + Send,
+{
+    type SampleMaybeUninit<'a>: SampleMaybeUninit<T> + 'a
+    where
+        Self: 'a;
+
+    fn allocate<'a>(&'a self) -> Result<Self::SampleMaybeUninit<'a>>;
+
+    fn send(&self, value: T) -> Result<()> {
+        let sample = self.allocate()?;
+        let init_sample = sample.write(value);
+        init_sample.send()
+    }
+}
+
+pub trait Consumer<R: Runtime + ?Sized> {
+    fn new(instance_info: R::ConsumerInfo) -> Self;
```

can that fail ?

---

i don't think so this can fail because it is just creating instance. 
But if service not available at the time of consumer creation or after that then user will not able to receive data.
So we have both possibility where to send error either at the receiving APIs or at consumer creation.
As per my understanding we should have at the receiving APIs side.

---

Can this fail due to missing resources (os allocations, issues in file systems)? In iceoryx2 it can fail imho (just code wise.)

---

You’re right that new() often looks like a simple constructor, but in practice it can fail if the implementation needs to acquire resources or validate runtime state. For example, OS-level allocations, shared memory mapping, or file system access can all fail.

From an API design perspective, we have two options:

- Fail early in new() if required resources or invariants aren’t available.
- Defer errors to receive() if we allow lazy initialization.

We need design decision here - @darkwisebear , @lurtz , @pawelrutkaq

---

No strong opinion on that. Making it fallible makes the life easier for the implementor of the backend, but the user has to cover yet another error path - however, since this is during initialization, there's anyway a lot of error handling going on. I wouldn't object making this a fallible API.

---

Then lets do failable please ;) indeed will make our life easier (both publisher and subscriber)